Hover state lost when moving from input[type=checkbox] to parent label and back again

Ahmad-S792 opened the pull request:
#### 13c06357a7d55e6bfbf9c5327f5aa38658926f05
<pre>
Hover state lost when moving from input[type=checkbox] to parent label and back again
<a href="https://bugs.webkit.org/show_bug.cgi?id=254110">https://bugs.webkit.org/show_bug.cgi?id=254110</a>
<a href="https://rdar.apple.com/107220591">rdar://107220591</a>

Reviewed by Darin Adler.

This patch aligns WebKit with Gecko / Firefox and Blink / Chromium.

Merge: <a href="https://github.com/chromium/chromium/commit/90f27accb21eafe9e237b454905bf6c5a590e0ab">https://github.com/chromium/chromium/commit/90f27accb21eafe9e237b454905bf6c5a590e0ab</a>

Before this patch if the commonAncestor of newHovered and oldHovered
node is newHovered node then setHover gets ignored for newHovered node.

We are not merging test because it is already imported in past and passing somehow,
although I manually tested the new behavior on `<a href="https://jsfiddle.net/MT9kL/`">https://jsfiddle.net/MT9kL/`</a> and confirm
that our behavior now matches with other browsers on above JSFiddle.

* Source/WebCore/dom/Document.cpp:
(WebCore::Document::updateHoverActiveState):
* Source/WebCore/html/HTMLLabelElement.cpp:
(WebCore::HTMLLabelElement::setActive):
(WebCore::HTMLLabelElement::setHovered):

Canonical link: <a href="https://commits.webkit.org/288449@main">https://commits.webkit.org/288449@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8bc871f3b977fc3d8343e60b36ed83f5d360c274

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/82354 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/2018 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/36433 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/87486 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/33415 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/84459 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/2089 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/9928 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/64184 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/21935 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/85423 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/1471 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/74934 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/44462 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/1372 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/29117 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/32456 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/72663 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/29745 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/88842 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/9660 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/6944 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/72585 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/9886 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/70748 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/71802 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17957 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/15938 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/14946 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/1019 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/9613 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/15134 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/9487 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/12953 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/11257 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder reviewed (commented):
Code changes look fine but a couple of comments:

1. I think we should retitle the bug / commit to "Hover state lost when moving from input[type=checkbox] to parent label". From testing on the JSFiddle, the hover state is present when moving "back again".

2. The reason the existing imported test is already passing is because `box-shadow` doesn't apply to checkboxes with native appearance. This means that the reference trivially matches. We should move the test to `fast/forms/label` and rewrite it to use `appearance: none`, give the checkbox a width/height, and change the border/background on hover. If done correctly, the test should fail before this change, and pass after it.

Ahmad-S792 committed:
Hover state lost when moving from input[type=checkbox] to parent label and back again
https://bugs.webkit.org/show_bug.cgi?id=254110
rdar://107220591

Reviewed by Darin Adler.

This patch aligns WebKit with Gecko / Firefox and Blink / Chromium.

Merge: https://github.com/chromium/chromium/commit/90f27accb21eafe9e237b454905bf6c5a590e0ab

Before this patch if the commonAncestor of newHovered and oldHovered
node is newHovered node then setHover gets ignored for newHovered node.

We are not merging test because it is already imported in past and passing somehow,
although I manually tested the new behavior on `https://jsfiddle.net/MT9kL/` and confirm
that our behavior now matches with other browsers on above JSFiddle.

* Source/WebCore/dom/Document.cpp:
(WebCore::Document::updateHoverActiveState):
* Source/WebCore/html/HTMLLabelElement.cpp:
(WebCore::HTMLLabelElement::setActive):
(WebCore::HTMLLabelElement::setHovered):

Canonical link: https://commits.webkit.org/288449@main
